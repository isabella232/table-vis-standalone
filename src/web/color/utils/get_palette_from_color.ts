import { hsl } from 'd3-color'

/**
 * This function returns a generated color palette based on a base color.
 * The Lightness values are derived-estimates from the Lens color-shade distributions.
 */
export const getPaletteFromColor = (color: string) => {
  const { h, s } = hsl(color)
  return {
    color000: hsl(h, s, 0.97).hex(),
    color100: hsl(h, s, 0.95).hex(),
    color200: hsl(h, s, 0.9).hex(),
    color300: hsl(h, s, 0.75).hex(),
    color400: hsl(h, s, 0.6).hex(),
    color500: hsl(h, s, 0.45).hex(),
    color600: hsl(h, s, 0.35).hex(),
    color700: hsl(h, s, 0.3).hex(),
    color800: hsl(h, s, 0.23).hex(),
    color900: hsl(h, s, 0.17).hex(),
  } as ColorPalette
}

export interface ColorPalette {
  color000: string
  color100: string
  color200: string
  color300: string
  color400: string
  color500: string
  color600: string
  color700: string
  color800: string
  color900: string
}
