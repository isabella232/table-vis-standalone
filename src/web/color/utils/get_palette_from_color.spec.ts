import { getPaletteFromColor } from './get_palette_from_color'

describe('getPaletteFromColor', () => {
  it('should return a palette', () => {
    expect(getPaletteFromColor('white')).toMatchSnapshot('white')
    expect(getPaletteFromColor('purple')).toMatchSnapshot('purple')
  })

  it('should accept invalid colors and return default palette', () => {
    expect(getPaletteFromColor('not a color')).toMatchSnapshot()
    expect(getPaletteFromColor('')).toMatchSnapshot()
  })
})
