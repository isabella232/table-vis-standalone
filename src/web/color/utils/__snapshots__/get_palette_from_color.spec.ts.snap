// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPaletteFromColor should accept invalid colors and return default palette 1`] = `
Object {
  "color000": "#f7f7f7",
  "color100": "#f2f2f2",
  "color200": "#e6e6e6",
  "color300": "#bfbfbf",
  "color400": "#999999",
  "color500": "#737373",
  "color600": "#595959",
  "color700": "#4d4d4d",
  "color800": "#3b3b3b",
  "color900": "#2b2b2b",
}
`;

exports[`getPaletteFromColor should accept invalid colors and return default palette 2`] = `
Object {
  "color000": "#f7f7f7",
  "color100": "#f2f2f2",
  "color200": "#e6e6e6",
  "color300": "#bfbfbf",
  "color400": "#999999",
  "color500": "#737373",
  "color600": "#595959",
  "color700": "#4d4d4d",
  "color800": "#3b3b3b",
  "color900": "#2b2b2b",
}
`;

exports[`getPaletteFromColor should return a palette: purple 1`] = `
Object {
  "color000": "#fff0ff",
  "color100": "#ffe5ff",
  "color200": "#ffccff",
  "color300": "#ff80ff",
  "color400": "#ff33ff",
  "color500": "#e600e6",
  "color600": "#b300b3",
  "color700": "#990099",
  "color800": "#750075",
  "color900": "#570057",
}
`;

exports[`getPaletteFromColor should return a palette: white 1`] = `
Object {
  "color000": "#f7f7f7",
  "color100": "#f2f2f2",
  "color200": "#e6e6e6",
  "color300": "#bfbfbf",
  "color400": "#999999",
  "color500": "#737373",
  "color600": "#595959",
  "color700": "#4d4d4d",
  "color800": "#3b3b3b",
  "color900": "#2b2b2b",
}
`;
